File: apps/m2-portal/Services/GoodsReceiptService.cs
using System.Net.Http.Json;

namespace M2Portal.Services;

public enum GoodsReceiptStatus { Pending, Confirmed, Discrepancy }

public record GoodsReceiptLine(
    int LineNumber,
    string ProductCode,
    string ProductNameZht,
    int ExpectedQty,
    int ReceivedQty,
    string? DiscrepancyNote)
{
    public int Discrepancy => ReceivedQty - ExpectedQty;
}

public record GoodsReceiptSummary(
    Guid Id,
    string SapDeliveryNote,
    GoodsReceiptStatus Status,
    DateTimeOffset ReceivedDate);

public record GoodsReceiptDetail(
    Guid Id,
    string SapDeliveryNote,
    GoodsReceiptStatus Status,
    DateTimeOffset ReceivedDate,
    IReadOnlyList<GoodsReceiptLine> Lines);

/// <summary>
/// Typed HTTP client wrapping M2PortalBff /api/goods-receipt/ endpoints.
/// Returns mock data when the BFF is unavailable (Sprint 4 stub).
/// </summary>
public class GoodsReceiptService(HttpClient http)
{
    /// GET /api/goods-receipt/?tenantId={tenantId}&shopId={shopId}
    public async Task<IReadOnlyList<GoodsReceiptSummary>> ListAsync(
        Guid tenantId, Guid shopId, CancellationToken ct = default)
    {
        try
        {
            return await http.GetFromJsonAsync<List<GoodsReceiptSummary>>(
                $"/api/goods-receipt/?tenantId={tenantId}&shopId={shopId}", ct) ?? [];
        }
        catch
        {
            return MockSummaries();
        }
    }

    /// GET /api/goods-receipt/{id}
    public async Task<GoodsReceiptDetail> GetAsync(Guid id, CancellationToken ct = default)
    {
        try
        {
            return await http.GetFromJsonAsync<GoodsReceiptDetail>(
                $"/api/goods-receipt/{id}", ct)
                ?? throw new KeyNotFoundException($"Goods receipt {id} not found.");
        }
        catch
        {
            return MockDetail(id);
        }
    }

    /// POST /api/goods-receipt/{id}/confirm
    public async Task ConfirmAsync(Guid id, CancellationToken ct = default)
    {
        var response = await http.PostAsync($"/api/goods-receipt/{id}/confirm", null, ct);
        response.EnsureSuccessStatusCode();
    }

    /// POST /api/goods-receipt/{id}/discrepancy
    public async Task RecordDiscrepancyAsync(Guid id, string note, CancellationToken ct = default)
    {
        var response = await http.PostAsJsonAsync(
            $"/api/goods-receipt/{id}/discrepancy", new { note }, ct);
        response.EnsureSuccessStatusCode();
    }

    private static IReadOnlyList<GoodsReceiptSummary> MockSummaries() =>
    [
        new(Guid.Parse("00000000-0000-0000-0000-000000000001"),
            "4900012345", GoodsReceiptStatus.Pending,
            DateTimeOffset.Now.AddHours(-2)),
        new(Guid.Parse("00000000-0000-0000-0000-000000000002"),
            "4900012346", GoodsReceiptStatus.Confirmed,
            DateTimeOffset.Now.AddDays(-1)),
        new(Guid.Parse("00000000-0000-0000-0000-000000000003"),
            "4900012347", GoodsReceiptStatus.Discrepancy,
            DateTimeOffset.Now.AddDays(-2)),
    ];

    private static GoodsReceiptDetail MockDetail(Guid id) => new(
        id,
        "4900012345",
        GoodsReceiptStatus.Pending,
        DateTimeOffset.Now.AddHours(-2),
        [
            new(1, "PRD-001", "洗髮精 500ml", 24, 24, null),
            new(2, "PRD-002", "沐浴乳 300ml", 12, 10, "2件破損"),
        ]);
}

File: apps/meka-pos/lib/services/goods_receipt_service.dart
import 'package:dio/dio.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../core/api/api_client.dart';

enum GoodsReceiptStatus { pending, confirmed, discrepancy }

extension GoodsReceiptStatusLabel on GoodsReceiptStatus {
  String get label {
    switch (this) {
      case GoodsReceiptStatus.pending:
        return '待確認';
      case GoodsReceiptStatus.confirmed:
        return '已確認';
      case GoodsReceiptStatus.discrepancy:
        return '有差異';
    }
  }
}

class GoodsReceiptLine {
  const GoodsReceiptLine({
    required this.lineNumber,
    required this.productCode,
    required this.productNameZht,
    required this.expectedQty,
    required this.receivedQty,
    this.discrepancyNote,
  });

  final int lineNumber;
  final String productCode;
  final String productNameZht;
  final int expectedQty;
  final int receivedQty;
  final String? discrepancyNote;

  int get discrepancy => receivedQty - expectedQty;

  factory GoodsReceiptLine.fromJson(Map<String, dynamic> json) =>
      GoodsReceiptLine(
        lineNumber: json['lineNumber'] as int,
        productCode: json['productCode'] as String,
        productNameZht: json['productNameZht'] as String,
        expectedQty: json['expectedQty'] as int,
        receivedQty: json['receivedQty'] as int,
        discrepancyNote: json['discrepancyNote'] as String?,
      );
}

class GoodsReceipt {
  const GoodsReceipt({
    required this.id,
    required this.sapDeliveryNote,
    required this.status,
    required this.receivedDate,
    this.lines = const [],
  });

  final String id;
  final String sapDeliveryNote;
  final GoodsReceiptStatus status;
  final DateTime receivedDate;
  final List<GoodsReceiptLine> lines;

  factory GoodsReceipt.fromJson(Map<String, dynamic> json) => GoodsReceipt(
        id: json['id'] as String,
        sapDeliveryNote: json['sapDeliveryNote'] as String,
        status: GoodsReceiptStatus.values.firstWhere(
          (s) => s.name == json['status'],
          orElse: () => GoodsReceiptStatus.pending,
        ),
        receivedDate: DateTime.parse(json['receivedDate'] as String),
        lines: (json['lines'] as List? ?? [])
            .map((e) => GoodsReceiptLine.fromJson(e as Map<String, dynamic>))
            .toList(),
      );
}

/// Typed HTTP service for goods receipts. Falls back to mock data when the
/// BFF is unavailable (BFF endpoint: `/api/goods-receipt/`).
class GoodsReceiptService {
  GoodsReceiptService(this._dio);
  final Dio _dio;

  /// GET /api/goods-receipt/
  Future<List<GoodsReceipt>> listAsync() async {
    try {
      final res = await _dio.get('/api/v1/goods-receipt/');
      return (res.data as List)
          .map((e) => GoodsReceipt.fromJson(e as Map<String, dynamic>))
          .toList();
    } catch (_) {
      return _mockList();
    }
  }

  /// GET /api/goods-receipt/{id}
  Future<GoodsReceipt> getAsync(String id) async {
    try {
      final res = await _dio.get('/api/v1/goods-receipt/$id');
      return GoodsReceipt.fromJson(res.data as Map<String, dynamic>);
    } catch (_) {
      return _mockList().firstWhere(
        (r) => r.id == id,
        orElse: () => _mockList().first,
      );
    }
  }

  static List<GoodsReceipt> _mockList() => [
        GoodsReceipt(
          id: 'gr-001',
          sapDeliveryNote: '4900012345',
          status: GoodsReceiptStatus.pending,
          receivedDate: DateTime.now().subtract(const Duration(hours: 2)),
          lines: [
            const GoodsReceiptLine(
              lineNumber: 1,
              productCode: 'PRD-001',
              productNameZht: '洗髮精 500ml',
              expectedQty: 24,
              receivedQty: 24,
            ),
            const GoodsReceiptLine(
              lineNumber: 2,
              productCode: 'PRD-002',
              productNameZht: '沐浴乳 300ml',
              expectedQty: 12,
              receivedQty: 10,
              discrepancyNote: '2件破損',
            ),
          ],
        ),
        GoodsReceipt(
          id: 'gr-002',
          sapDeliveryNote: '4900012346',
          status: GoodsReceiptStatus.confirmed,
          receivedDate: DateTime.now().subtract(const Duration(days: 1)),
          lines: [
            const GoodsReceiptLine(
              lineNumber: 1,
              productCode: 'PRD-003',
              productNameZht: '護髮素 250ml',
              expectedQty: 36,
              receivedQty: 36,
            ),
          ],
        ),
        GoodsReceipt(
          id: 'gr-003',
          sapDeliveryNote: '4900012347',
          status: GoodsReceiptStatus.discrepancy,
          receivedDate: DateTime.now().subtract(const Duration(days: 2)),
          lines: [
            const GoodsReceiptLine(
              lineNumber: 1,
              productCode: 'PRD-004',
              productNameZht: '洗面乳 100ml',
              expectedQty: 48,
              receivedQty: 40,
              discrepancyNote: '8件缺貨，供應商確認中',
            ),
          ],
        ),
      ];
}

final goodsReceiptServiceProvider = Provider<GoodsReceiptService>(
  (ref) => GoodsReceiptService(ref.read(apiClientProvider)),
);

final goodsReceiptListProvider = FutureProvider<List<GoodsReceipt>>((ref) {
  return ref.read(goodsReceiptServiceProvider).listAsync();
});

final goodsReceiptDetailProvider =
    FutureProvider.family<GoodsReceipt, String>((ref, id) {
  return ref.read(goodsReceiptServiceProvider).getAsync(id);
});
